velox/expression/CastExpr.cpp

```diff
@@ -241,6 +241,45 @@ void applyIntToDecimalCastKernel(
   });
 }
 
+template <
+    typename TInput,
+    typename TOutput,
+    typename = typename std::enable_if_t<std::is_floating_point_v<TInput>>>
+void applyDoubleToDecimal(
+    const SelectivityVector& rows,
+    const BaseVector& input,
+    exec::EvalCtx& context,
+    const TypePtr& toType,
+    VectorPtr& castResult) {
+  auto sourceVector = input.as<SimpleVector<TInput>>();
+  auto castResultRawBuffer =
+      castResult->asUnchecked<FlatVector<TOutput>>()->mutableRawValues();
+  const auto& toPrecisionScale = getDecimalPrecisionScale(*toType);
+  context.applyToSelectedNoThrow(rows, [&](vector_size_t row) {
+    const auto& sourceValue = sourceVector->valueAt(row);
+    if (!std::isfinite(sourceValue)) {
+      castResult->setNull(row, true);
+      return;
+    }
+
+    auto toValue = sourceValue *
+        static_cast<TInput>(DecimalUtil::kPowersOfTen[toPrecisionScale.second]);
+
+    if (toValue >= 0) {
+      toValue = std::floor(toValue + 0.5);
+    } else {
+      toValue = std::ceil(toValue - 0.5);
+    }
+
+    if (toValue <= static_cast<TInput>(std::numeric_limits<TOutput>::min()) ||
```

Comment:
And you should check the range by `DecimalUtil`
```
  FOLLY_ALWAYS_INLINE static void valueInRange(int128_t value) {
    VELOX_CHECK(
        (value >= kLongDecimalMin && value <= kLongDecimalMax),
        "Value '{}' is not in the range of Decimal Type",
        value);
  }
```

Reply:
VELOX_CHECK will throw an exception if the result of the expression is false. So that means, we should throw an exception here instead of setting a null value?

Reply:
Yes, like this exception https://github.com/facebookincubator/velox/blob/main/velox/type/DecimalUtil.h#L141

In Spark, it will set null instead of throwing exception, but now the cast behavior consists with Presto